### PR TITLE
Candidate for version 4.0.17

### PR DESCRIPTION
**Pull request by kushti**
* #1332
* #1464
* #1483
* #1492
* #1493 
* #1494
* #1497 
* #1499
* #1501
* #1507 
* #1508
* #1512
* #1515
* #1522
* #1530
* #1537
* #1538

Contributors: @ApexTheory , @ccellado, @chriswill, @knizhnik , @kushti , @MrStahlfelge, @pragmaxim ,  @taikatytto

### CODE REVIEW

**Comment by glasgowm148** [comment]
Tested & Working on M1 & Pi

File: src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala [hunk 1]
```diff
@@ -723,6 +721,9 @@ object CandidateGenerator extends ScorexLogging {
                     val boxesToSpend = feeTx.inputs.flatMap(i =>
                       newBoxes.find(b => java.util.Arrays.equals(b.id, i.boxId))
                     )
+                    implicit val verifier: ErgoInterpreter = ErgoInterpreter(
+                      us.stateContext.currentParameters
+                    )
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala:726]: I remember `implicit` was removed and the verifier was moved out of the loop. And also resetContext was used. Something changed?

File: src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala [hunk 2]
```diff
@@ -71,7 +71,7 @@ case class ErgoTransaction(override val inputs: IndexedSeq[Input],
     Algos.hash(ByteArrayUtils.mergeByteArrays(inputs.map(_.spendingProof.proof))).tail
 
 
-  lazy val outAssetsTry: Try[(Map[ByteArrayWrapper, Long], Int)] = ErgoTransaction.extractAssets(outputCandidates)
+  lazy val outAssetsTry: Try[(Map[ByteBuffer, Long], Int)] = ErgoBoxAssetExtractor.extractAssets(outputCandidates)
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala:74]: What is the reason for this radical change in the key type?
 ByteArrayWrapper and ByteBuffer have completely different equality. 
ByteBuffer's equality is position sensitive, i.e. any reading from it will change the key.
Kind of scary to have such object as a key in the Map.

File: src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala [hunk 3]
```diff
@@ -185,6 +184,35 @@ trait ErgoHistory
...
+      ).get
+      requiredModifiersForHeader(h).foreach { case (_, mId) =>
+        historyStorage.remove(
+          Seq(
+            validityKey(mId)
+          ),
+          Seq(
+            mId
+          )
+        ).get
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala:212]: Kind of weird formatting :). Let's maybe use named arguments for clarity.

File: src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/FullBlockProcessor.scala [hunk 4]
```diff
@@ -232,7 +232,7 @@ trait FullBlockProcessor extends HeadersProcessor {
     val toRemove: Seq[ModifierId] = heights.flatMap(h => headerIdsAtHeight(h))
       .flatMap(id => typedModifierById[Header](id))
       .flatMap(_.sectionIds.map(_._2))
-    historyStorage.remove(toRemove)
+    historyStorage.remove(Seq.empty, toRemove)
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/FullBlockProcessor.scala:235]: Just a performance hint. The fastest way to obtain empty Seq is to use `mutable.WrappedArray.empty[T]`

File: ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
+package org.ergoplatform.wallet.interface4j;
```
Comment (aslesarenko) [inline ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java:1]: Is this copied from Appkit or it has some difference? Please comment in the code with direct reference if necessary.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: src/main/resources/api/openapi.yaml [hunk 6]
```diff
@@ -1526,7 +1540,7 @@ components:
     PeersStatus:
       type: object
       required:
-        - lastInomingMessage
+        - lastIncomingMessage
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/resources/api/openapi.yaml:1543]: No, only OpenAPI description of the API changed,  not API itself

File: src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala (hunk 1 above)
Comment (pragmaxim) [reply src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala:726]: I was also thinking why it cannot be outside of the loop, initialized only once

File: src/main/resources/api/openapi.yaml [hunk 7]
```diff
@@ -3600,7 +3614,7 @@ paths:
                 $ref: '#/components/schemas/ApiError'
 
   /wallet/boxes/collect:
-    get:
+    post:
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/resources/api/openapi.yaml:3617]: Only OpenAPI description of the API changed,  not API itself

File: ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala [hunk 8]
```diff
@@ -17,7 +17,7 @@ object ArithUtils {
     */
   @inline def multiplyExact(e1: Long, e2: Long): Long = {
     try {
-      Math.multiplyExact(e1, e2)
+      java7.compat.Math.multiplyExact(e1, e2)
```
[the remark this answers is not in the document]
Comment (MrStahlfelge) [reply ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala:20]: It's the one from sigmastate interpreter added here https://github.com/ScorexFoundation/sigmastate-interpreter/pull/736

File: src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala (hunk 1 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala:726]: That is done in another PR https://github.com/ergoplatform/ergo/pull/1522

File: src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala (hunk 2 above)
Comment (pragmaxim) [reply src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala:74]: uff, nice catch ... the ByteArrayWrapper is obsolete and we wanted to get rid of it so an alternative was needed.

Comment (pragmaxim) [reply src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala:74]: I will go with akka's `ByteString` if you agree

**Review by greenhat — APPROVED** [review]

File: ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala:20]: Oh, I should have made it more clear. 
The whole this method was just brought to my eyes by this change.

I'm talking about catching exceptions and returning Long.MaxValue.
What about negative overflow?
/cc @kushti

[the remark this answers is not in the document]
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala:20]: This method is used for cost calculation, so likely no subsequent problems.
But I think this should be fixed.

File: src/main/resources/api/openapi.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/resources/api/openapi.yaml:1543]: ah, so it was changed to reflect the actual implementation?

File: src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala (hunk 2 above)
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala]: the extractAssets method has nothing to do with akka dependency.
It actually can be moved in ergo-wallet module as utility method.
I suggest to use `Seq[Byte]` as key here and put ids there in WrappedArray.
val aiWrapped = WrappedArray.ofByte(assetId)

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/ErgoBoxAssetExtractor.scala [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    * @return total assets access cost
+    */
+  def totalAssetsAccessCost(
+    inAssetsNum: Int,
+    inAssetsSize: Int,
+    outAssetsNum: Int,
+    outAssetsSize: Int,
+    tokenAccessCost: Int
+  ): Int =
+    (outAssetsNum + inAssetsNum) * tokenAccessCost + (inAssetsSize + outAssetsSize) * tokenAccessCost
```
Comment (aslesarenko) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/ErgoBoxAssetExtractor.scala:61]: Cost accumulation should always throw on arithmetic overflow.

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/ErgoBoxAssetExtractor.scala [hunk 10]
```diff
@@ -0,0 +1,89 @@
+package org.ergoplatform.wallet.boxes
+
+import akka.util.ByteString
```
Comment (aslesarenko) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/ErgoBoxAssetExtractor.scala:3]: I suggest to avoid akka dependency. See my other comment

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: src/main/resources/api/openapi.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/resources/api/openapi.yaml:1543]: Yes

File: ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala:20]: @aslesarenko if you think behavior before refactoring has problems, please make an issue. For the scope of this PR I think equivalence should be required only. To ensure it, a test can be done @MrStahlfelge where old code is checked against new one on edge cases (and maybe on random samples as well)

File: src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala (hunk 2 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala]: @aslesarenko why not to follow https://github.com/ergoplatform/ergo/issues/1132 instead ?  With your proposal it seems we're getting more and more wrappers for fixed-size byte arrays

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/ErgoBoxAssetExtractor.scala (hunk 9 above)
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/ErgoBoxAssetExtractor.scala:61]: @pragmaxim please use addExact / multiplyExact here

Comment (MrStahlfelge) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/ErgoBoxAssetExtractor.scala:61]: please use it from `java7.compat.Math` class since this is in wallet module and is used in appkit.

File: ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala:20]: @kushti the `java7.compat.Math` methods are being tested against `Math` and test vectors methods and test vectors in Sigma as they are used there for all exact operations

File: src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala (hunk 2 above)
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala]: @kushti, @pragmaxim  yes, but #1132 is not implemented yet. The proposed pattern can be implemented though under a different name (WrappedHash, HashBytes, HashKey, etc.) to not confuse with ModifierId.
It can be used everywhere a hash is used for better safety and performance.

File: ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+import java.util.Arrays;
+
+/**
+ * Encapsulates secret array of characters (char[]) with proper equality.
+ * The secret data can be {@link SecretString#erase() erased} in memory and not leaked to GC.
+ * Using this class is more secure and safe than using char[] directly.
+ */
+public final class SecretString {
```
[the remark this answers is not in the document]
Comment (ccellado) [reply ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java:10]: Agreed, should add this separately.

File: ergo-wallet/src/test/scala/org/ergoplatform/wallet/mnemonic/MnemonicSpec.scala [hunk 12]
```diff
@@ -402,8 +403,8 @@ class MnemonicSpec
     val sentenceSize = if (langId == "japanese") sentence.split("\u3000").length else sentence.split(" ").length
     val strength = Mnemonic.AllowedStrengths.zip(Mnemonic.MnemonicSentenceSizes).find(_._2 == sentenceSize).map(_._1).get
     val mnemonic = new Mnemonic(langId, strength)
-    Base16.encode(Mnemonic.toSeed(sentence, Some(pass))) shouldEqual seed
-    normalize(mnemonic.toMnemonic(Base16.decode(entropy).get).get, NFKD) shouldEqual normalize(sentence, NFKD)
+    Base16.encode(Mnemonic.toSeed(SecretString.create(sentence), Some(SecretString.create(pass)))) shouldEqual seed
```
[the remark this answers is not in the document]
Comment (ccellado) [reply ergo-wallet/src/test/scala/org/ergoplatform/wallet/mnemonic/MnemonicSpec.scala:406]: Ok

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/SecretStorage.scala [hunk 13]
```diff
@@ -11,10 +12,10 @@ trait SecretStorage {
 
   def secret: Option[ExtendedSecretKey]
 
-  def unlock(pass: String): Try[Unit]
+  def unlock(pass: SecretString): Try[Unit]
 
   def lock(): Unit
 
-  def checkSeed(mnemonic: String, mnemonicPassOpt: Option[String]): Boolean
+  def checkSeed(mnemonic: SecretString, mnemonicPassOpt: Option[SecretString]): Boolean
```
[the remark this answers is not in the document]
Comment (ccellado) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/SecretStorage.scala:19]: Ok

File: ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java (hunk 5 above)
Comment (ccellado) [reply ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java:1]: Noted

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/SecretStorage.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (ccellado) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/SecretStorage.scala:19]: Ok

**Comment by ccellado** [comment]
Created new mr #1538 against 4.0.18 for SecretString for a separate release.

@kushti please remove #1497 , as discussed with @aslesarenko.

File: ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply ergo-wallet/src/main/scala/org/ergoplatform/utils/ArithUtils.scala:20]: created https://github.com/ergoplatform/ergo/issues/1535

File: src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala (hunk 3 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala:202]: fixed

File: ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (kushti) [reply ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java:10]: @aslesarenko @ccellado likely a separate release with SecretString just will be skipped just. So I propose to do functional testing for 4.0.17 with the SecretString PR included

**Review by pragmaxim — APPROVED** [review]

File: ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java:10]: An alternative is to update SecretString so that after it is erased, any further method calls on it will throw an exception.

[the remark this answers is not in the document]
Comment (ccellado) [reply ergo-wallet/src/main/java/org/ergoplatform/wallet/interface4j/SecretString.java:10]: That sounds reasonable. I'll implement that

**Review by aslesarenko — COMMENTED** [review]
LGTM, but, there is a lot of changes, I wonder if all of them covered with tests?

**Review by aslesarenko — APPROVED** [review]

**Comment by kushti** [comment]
> LGTM, but, there is a lot of changes, I wonder if all of them covered with tests?

No, but for ErgoNodeViewSynchronizerSpecification first few tests appeared very recently, and we're currently going for refactoring , specification and then doing more tests
